src/haliax/_src/state_dict.py

```diff
@@ -0,0 +1,491 @@
+                pspec = partitioning.pspec_for_axis(tree.axes)
+                sharding = jax.sharding.NamedSharding(mesh, pspec)
+                array = jax.make_array_from_callback(tree.array.shape, sharding, lambda indices: array[indices])
+            else:
+                array = jnp.array(array)
+            array = named(array, tree.axes)
+        else:
+            array = named(array, tree.axes)
+            array = partitioning.auto_sharded(array)
+
+        return array
+    elif is_jax_array_like(tree):
+        if prefix is None:
+            raise ValueError("Cannot extract a leaf value from a state dict without a prefix")
+        # TODO: add "strict" flag so we can return None in cases where it's just missing
+        return jnp.array(state_dict[prefix])
+    else:
+        if prefix is None:
+            return tree
+        return state_dict.get(prefix, tree)
+
+
+def update_state_dict(tree: PyTree, state_dict: StateDict, prefix: Optional[str] = None) -> None:
+    """
+    Update a state dict with the values from a tree.
+
+    !!! warning
+        This function modifies the state dict in place.
+
+    Args:
+        tree:  The tree to update the state dict with.
+        state_dict:  The state dict to update.
+        prefix:  The prefix to use when updating the state dict.
+
+    """
+    if isinstance(tree, eqx.Module):
+        if hasattr(tree, "update_state_dict"):
+            tree.update_state_dict(state_dict, prefix)
+        else:
+            default_update_state_dict_with_eqx_module(state_dict, tree, prefix)
+    elif isinstance(tree, list):
+        for i, item in enumerate(tree):
+            update_state_dict(item, state_dict, prefix=apply_prefix(prefix, str(i)))
+    elif isinstance(tree, dict):
+        for k, v in tree.items():
+            update_state_dict(v, state_dict, prefix=apply_prefix(prefix, k))
+    elif isinstance(tree, NamedArray):
+        assert prefix is not None
+        state_dict[prefix] = tree.array
+    elif is_jax_array_like(tree):
+        if prefix is not None:
+            if tree is not None:
+                state_dict[prefix] = tree  # type: ignore
+        else:
+            raise ValueError("Cannot update state dict with a leaf value.")
+    else:
+        pass
+
+
+def to_state_dict(tree: PyTree, prefix: Optional[str] = None) -> StateDict:
+    """
+    Convert a PyTree to a state dict.
+
+    Returns:
+        The state dict representation of the input tree.
+    """
+    state_dict: StateDict = {}
+    update_state_dict(tree, state_dict, prefix)
+    return state_dict
+
+
+def default_eqx_module_from_state_dict(mod: Mod, state_dict: StateDict, prefix: Optional[str] = None) -> Mod:
+    key_map: Dict[str, Optional[str]] = getattr(mod, "_state_dict_key_map", lambda: {})()  # type: ignore
+    names = []
+    values = []
+    for field in dataclasses.fields(mod):
+        if field.metadata.get("static", False):
```

Comment:
This is just me being dumb, but I read this backwards as "only process static fields" at first. Maybe just a comment for us dummies:

"# Filter out the dynamic fields for encoding in the `state_dict`"

(And IIUC, static fields are then handled when you construct the initial model)

Reply:
ah ok i'll add a comment